File: Python/DFS/SymmetricTree.py
# @author
# Aakash Verma

# Output:
# True


from collections import deque

class Node:
    def __init__(self, data):
        self.data = data
        self.right = self.left = None

class SymmetricTree:
    
    def __init__(self): 
        self.root = None

    def findSymmetric(self, root1, root2):
        if root1 is None and root2 is None:
            return True
        if root1 is None or root2 is None:
            return False
        return (root1.data == root2.data) and self.findSymmetric(root1.left, root2.right) and self.findSymmetric(root1.right, root2.left)

    def isSymmetric(self, root):
        
        return self.findSymmetric(root, root)

        
if __name__ == '__main__':
    tree = SymmetricTree()
    tree.root = Node(1)
    tree.root.left = Node(2)
    tree.root.right = Node(2)
    tree.root.left.left = Node(3)
    tree.root.left.right = Node(4)
    tree.root.right.left = Node(4)
    tree.root.right.right = Node(3)
    ans = tree.isSymmetric(tree.root)
    print(ans)
    

    

File: Java/DFS/SymmetricTree.java
/*
*   @author
*   Aakash Verma
*
*   Output:
*   true
*
*/

import java.util.*;

class Node {
    int data;
    Node left, right;
    public Node(int key) {
        data = key;
        left = right = null;
    }
}


class SymmetricTree {

    static Node root;
    SymmetricTree() {
        root = null;
    }

    public static boolean findSymmetric(Node root1, Node root2) {
        if(root1 == null && root2 == null) 
            return true;
        
        if(root1 == null || root2 == null) 
            return false;
        
        return (root1.data == root2.data) && 
                findSymmetric(root1.left, root2.right) && 
                findSymmetric(root1.right, root2.left);
    }

    public static boolean isSymmetric(Node root) {
        return findSymmetric(root, root);
    }

    public static void main (String[] args) {
        root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(2);
        root.left.left = new Node(3);
        root.left.right = new Node(4);
        root.right.left = new Node(4);
        root.right.right = new Node(3);
        boolean ans = isSymmetric(root);
        System.out.println(ans);

    }
}